utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,601 @@
+  toJSON     = String . vertexAttributeTypeToText
+  toEncoding = toEncoding . vertexAttributeTypeToText
+
+convertVertexAttributeType :: VertexAttributeType -> (Word32, Format)
+convertVertexAttributeType = \case
+  Vec2 -> (2 * floatSize, FORMAT_R32G32_SFLOAT)
+  Vec3 -> (3 * floatSize, FORMAT_R32G32B32_SFLOAT)
+  Vec4 -> (4 * floatSize, FORMAT_R32G32B32A32_SFLOAT)
+  where floatSize = 4
+
+data VertexAttribute = VertexAttribute
+  { binding  :: Word32
+  , location :: Word32
+  , size     :: Word32
+  , format   :: Format
+  }
+  deriving Show
+
+-- only support single binding for input vertex
+-- [SaschaWillems - Multiple Vulkan buffer binding points](https://gist.github.com/SaschaWillems/428d15ed4b5d71ead462bc63adffa93a)
+-- maybe can provide a binding map parameter to specify individual binding by name, like `foo [(1, ["inPos", "inColor"]) (2, ["texCoord"])]` speficies that `(binding = 1) inPos, (binding = 1) inColor, (binding = 2) texCoord`
+-- [island pipeline](https://github.com/tgfrerer/island/blob/76d0d38cba74181fa3774cef38aba4d96b6861dc/modules/le_backend_vk/le_pipeline.cpp#L21)
+makePipelineVertexInputStateCreateInfo
+  :: Vector Input -> Maybe (PipelineVertexInputStateCreateInfo '[])
+makePipelineVertexInputStateCreateInfo []     = Nothing
+makePipelineVertexInputStateCreateInfo inputs = Just zero
+  { vertexBindingDescriptions
+  , vertexAttributeDescriptions
+  }
+ where
+  vertexAttributes =
+    join . V.map makeVertexAttribute $ inputs :: Vector VertexAttribute
+  vertexAttributeDescriptions =
+    makeVertexInputAttributeDescriptions vertexAttributes :: Vector
+        VertexInputAttributeDescription
+  vertexBindingDescriptions =
+    makeVertexInputBindingDescriptions vertexAttributes :: Vector
+        VertexInputBindingDescription
+
+makeVertexInputBindingDescriptions
+  :: Vector VertexAttribute -> Vector VertexInputBindingDescription
+makeVertexInputBindingDescriptions =
+  V.fromList
+    . map (makeVertexInputBindingDescription . calculate)
+    . groupBy ((==) `on` fst)
+    . sortOn fst
+    . map extract
+    . V.toList
+ where
+  extract :: VertexAttribute -> ("binding" ::: Int, "size" ::: Int)
+  extract = liftA2 (,)
+                   (fromIntegral . (binding :: VertexAttribute -> Word32))
+                   (fromIntegral . size)
+  calculate
+    :: [("binding" ::: Int, "size" ::: Int)]
+    -> ("binding" ::: Int, "stride" ::: Int)
+  calculate = liftA2 (,) (fst . head) (sum . (snd <$>))
+
+makeVertexInputBindingDescription
+  :: ("binding" ::: Int, "stride" ::: Int) -> VertexInputBindingDescription
+makeVertexInputBindingDescription (binding, stride) = zero
+  { binding   = fromIntegral binding
+  , stride    = fromIntegral stride
+  , inputRate = VERTEX_INPUT_RATE_VERTEX
+  }
+
+makeVertexAttribute :: Input -> Vector VertexAttribute
+makeVertexAttribute Input {..} = V.generate
+  count
+  (\i -> VertexAttribute { binding  = 0
+                         , location = fromIntegral . (+ location) $ i
```

Comment:
`fromIntegral $ location + i`

That's why I avoid using `$` with a single final argument...

Reply:
fixed